### PR TITLE
Adds support for SSL proxy

### PR DESCRIPTION
**Pull request by jeffkole**
If you have server key and cert files, then you can have your proxy listen over
https as well as http.  This is very handy for when your app has features that
require SSL to work, such as accessing WebRTC streams.

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by rmg** [comment]
@slnode ok to test

File: proxy.js [hunk 1]
```diff
@@ -40,5 +45,26 @@ http.createServer(function (req, res) {
...
 });
+
+if (sslCert && sslKey) {
+  https.createServer({
+      key: fs.readFileSync(sslKey),
+      cert: fs.readFileSync(sslCert),
+      requestCert: false,
+      rejectUnauthorized: true
+    },
+    function (req, res) {
```
Comment (rmg) [inline proxy.js]: This function should probably be extracted and named, since it's the same for both cases.

**Comment by rmg** [comment]
@jeffkole thanks!

It's kind of unfortunate that there aren't any existing tests for proxy that can be extended to test this or verify that it didn't break anything :-(

I know it's a lot to ask, but do you happen to have any interest in adding basic tests for the proxy feature?

File: proxy.js [hunk 2]
```diff
@@ -40,5 +45,26 @@ http.createServer(function (req, res) {
...
     process.setuid(process.env.SUDO_USER);
   }
-
 });
+
+if (sslCert && sslKey) {
+  https.createServer({
+      key: fs.readFileSync(sslKey),
+      cert: fs.readFileSync(sslCert),
+      requestCert: false,
```
Comment (rmg) [inline proxy.js]: requestCert defaults to false, is this needed?

File: proxy.js [hunk 3]
```diff
@@ -40,5 +45,26 @@ http.createServer(function (req, res) {
...
+      rejectUnauthorized: true
```
Comment (rmg) [inline proxy.js]: This has no effect with `requestCert: false`, is it needed?

File: lib/proxy.js [hunk 4]
```diff
@@ -60,8 +69,28 @@ function startProxies(reqs, proc, command, emitter, portargs) {
...
     var ports = command.proxy.split(',');
+
+    var ssl = {
+      cert: '',
+      key:  ''
+    };
+    if(command.ssl) {
+      var certDirectoryContents = fs.readdirSync(command.ssl);
+      certDirectoryContents.forEach(function(file) {
+        if (file.indexOf('.crt') >= 0) {
```
Comment (rmg) [inline lib/proxy.js]: Is this intended to check file extensions only? If so `/\.crt$/.test(file)` would probably be a better conditional.

**Comment by rmg** [comment]
Thinking about it more.. as much as I like the simplicity of a single `--ssl` option, I think there are too many scenarios where it would do the wrong thing.

What do you think about splitting it into `--ssl-key` and `--ssl-cert` instead?

**Comment by jeffkole** [comment]
I'll take a crack at the tests as well as make changes based on your suggestions.  Thanks for the feedback.

**Comment by jeffkole** [comment]
Made the code updates you suggested (all good ones... thank you).  I'm going to work on the tests next.  And when the time is right, I'm happy to update the docs.  I assume that has to happen in a separate PR against the `gh-pages` branch though.

**Comment by rmg** [comment]
:+1: 

Let me know if you want to rebase this to rearrange the commits before I merge this.

**Comment by jeffkole** [comment]
Rebased.  Should be good to go.

**Pull request merged**